Publishing guide script

sharon-fdm opened the pull request:
# Checklist for submitter

If some of the following don't apply, delete the relevant line.

<!-- Note that API documentation changes are now addressed by the product design team. -->

- [ ] Changes file added for user-visible changes in `changes/` or `orbit/changes/`.
  See [Changes files](https://fleetdm.com/docs/contributing/committing-changes#changes-files) for more information.
- [ ] Input data is properly validated, `SELECT *` is avoided, SQL injection is prevented (using placeholders for values in statements)
- [ ] Added support on fleet's osquery simulator `cmd/osquery-perf` for new osquery data ingestion features.
- [ ] Added/updated tests
- [ ] If database migrations are included, checked table schema to confirm autoupdate 
- For database migrations:
  - [ ] Checked schema for all modified table for columns that will auto-update timestamps during migration.
  - [ ] Confirmed that updating the timestamps is acceptable, and will not cause unwanted side effects.
- [ ] Manual QA for all new/changed functionality
  - For Orbit and Fleet Desktop changes:
    - [ ] Manual QA must be performed in the three main OSs, macOS, Windows and Linux.
    - [ ] Auto-update manual QA, from released version of component to new version (see [tools/tuf/test](../tools/tuf/test/README.md)).

georgekarrv commented on tools/publish-automation/publish.go:
```diff
@@ -0,0 +1,269 @@
+func printIntro() {
+	fmt.Println("\n===========================================================")
+	fmt.Println("This Script is not full automation.\nIt will guide you through the publishing process step by step.")
+	fmt.Println("To start:\n- cd to your fleet root\n- go run ./tools/publish-automation/publish.go")
```
isn't this already running at this point?

georgekarrv commented on tools/publish-automation/publish.go:
```diff
@@ -0,0 +1,269 @@
+	fmt.Println("\n===========================================================")
+	fmt.Println("Step 1: Current version.")
+	for {
+		fmt.Println("Enter current version in the format similar to \"v4.47.2\": ")
```
current released version or current version you are trying to publish?

georgekarrv commented on tools/publish-automation/publish.go:
```diff
@@ -0,0 +1,269 @@
+	fmt.Println("Hitting ENTER will take you to the proper editing page (of current release)")
+	fmt.Println("Once published, go back here and hit ENTER again")
+	fmt.Scanln()
+	fmt.Println("Taking you to the edit page ... ")
```
we can release this in 1 line w/ `gh` I think that might be easier

lukeheath replied:
Agreed using the `gh` CLI would simplify things a bit.

georgekarrv replied:
https://github.com/fleetdm/fleet/pull/17870 added to patch_release script

georgekarrv commented on tools/publish-automation/publish.go:
```diff
@@ -0,0 +1,269 @@
+
+func openURL(url string) error {
+	var cmd *exec.Cmd
+
```
this wouldn't work for me 100% since I often have an ssh tunnel open to the machine I am releasing from

georgekarrv commented on tools/publish-automation/publish.go:
```diff
@@ -0,0 +1,269 @@
+	fmt.Println("\n===========================================================")
+	fmt.Println("Step 5: Check hub.docker for existing version")
+	fmt.Println("Hitting ENTER will take you to the proper hub.docker page.")
+	fmt.Println("Verify that the version exists")
```
The problem wasn't that the image didn't exist. You pushed the fleetctl image instead of the fleet server image.

When the image doesn't exist it fails more gracefullly

sharon-fdm replied:
Correct.
This will take the user to the proper image (And will give the correct image next step) so will prevent such mistakes.

georgekarrv replied:
again if we are running this script at the time this is ready to deploy we can achieve it in scripting w/ `gh`

georgekarrv replied:
My goal with scripts is that they are more than just a replica of documentation. Since maintaining both is more effort.

georgekarrv replied:
https://github.com/fleetdm/fleet/pull/17870 added into publish step of patch_release script

georgekarrv commented on tools/publish-automation/publish.go:
```diff
@@ -0,0 +1,269 @@
+	fmt.Println("Once at the git action page do this:")
+	fmt.Println("    1 - Go to #help-infrastructure slack channel and paste this:\"@infrastructure-oncall We have just released " + version + " and are now deploying it to dogfood\"  ")
+	fmt.Println("    2 - press the \"Run workflow\" button on the right side of the screen")
+	fmt.Println("    3 - paste \"fleetdm/fleet:patch-fleet-" + version + "\" as input.")
```
we have been avoiding using the `patch-fleet` image tag since having to deploy a second time fails if the image hasn't changed. Durring patch releases I always deploy the git tag

sharon-fdm replied:
That's how I release, but if we agree on tags, I don't mind.

lukeheath replied:
Using the `patch-fleet-` branch image is good for testing the RC, but at the end we should push the tag image: `fleet-v4.x.x`.

georgekarrv replied:
https://github.com/fleetdm/fleet/pull/17870 added to publish step of patch_release script

georgekarrv commented on tools/publish-automation/publish.go:
```diff
@@ -0,0 +1,269 @@
+	fmt.Println("Go to #help-engineering slack channel and paste this:")
+	fmt.Println("     We have just deployed " + version + " to dogfood")
+	fmt.Println("Go to #general slack channel and paste this:")
+	fmt.Println("     :cloud: :rocket: The latest version of Fleet is " + version[1:])
```
This is a copy from the slask `osquery/fleet` MOTD, this feels like needles duplication.

sharon-fdm replied:
Looks good to me to keep on both `osquery` and `fleetdm` but I'm open for changes.

sharon-fdm commented:
Thanks for the good comments @georgekarrv. 
I'll do a second round sometime soon. Hopefully get this ready for the next time I will patch.
Other than addressing your comments I may:
-  add auto slack notifications (If not too complicated)
- Fully automate less sensitive parts. (allowing review for sensitive ones)

lukeheath commented on tools/publish-automation/publish.go:
```diff
@@ -0,0 +1,268 @@
+	version := getVersion()
+	makeSureNPMVersionIsCoherent(version)
+	if !openReleasesPage(version) {
+		fmt.Println("Error - could not publish the fleat release.")
```
```suggestion
		fmt.Println("Error - could not publish the Fleet release.")
```

lukeheath reviewed (commented):
Cool! I don't think I will use it myself but I'm always happy to have more tools to assist with the release process.
